pkg/runit/servicebuilder.go

```diff
@@ -75,52 +109,160 @@ var DefaultBuilder = &ServiceBuilder{
+		// create the default log directory
+		err := os.MkdirAll(filepath.Join(stageDir, "log"), 0755)
+		if err != nil {
+			return err
+		}
+
+		runScript, err := template.RunScript()
+		if err != nil {
+			return err
+		}
+		if _, err := util.WriteIfChanged(filepath.Join(stageDir, "run"), runScript, 0755); err != nil {
+			return err
+		}
+
+		logScript, err := template.LogScript()
+		if err != nil {
+			return err
+		}
+		if _, err := util.WriteIfChanged(filepath.Join(stageDir, "log", "run"), logScript, 0755); err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
-	f, err := os.OpenFile(yamlPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
-	if err != nil {
-		return "", util.Errorf("Could not write servicebuilder template %s: %s", template.name, err)
+// symlink the runit service directory into the actual directory being monitored
+// by runsvdir
+// runsvdir will automatically start a service for each new directory (unless a
+// down file exists)
+func (s *ServiceBuilder) activate(templates map[string]ServiceTemplate) error {
+	for serviceName := range templates {
+		linkPath := filepath.Join(s.RunitRoot, serviceName)
+		stageDir := filepath.Join(s.StagingRoot, serviceName)
+
+		info, err := os.Lstat(linkPath)
+		if err == nil {
+			// if it exists, make sure it is actually a symlink
+			if info.Mode()&os.ModeSymlink == 0 {
+				return util.Errorf("%s is not a symlink", linkPath)
+			}
+
+			// and that it points to the right place
+			target, err := os.Readlink(linkPath)
+			if err != nil {
+				return err
+			}
+			if target != stageDir {
+				return util.Errorf("%s is a symlink to %s (expected %s)", linkPath, target, stageDir)
+			}
+		} else if os.IsNotExist(err) {
+			if err = os.Symlink(stageDir, linkPath); err != nil {
+				return err
+			}
+		} else if err != nil {
+			return err
+		}
 	}
-	defer f.Close()
+	return nil
+}
 
-	return yamlPath, template.Write(f)
+// public API to write, stage and activate a servicebuilder template
+func (s *ServiceBuilder) Activate(name string, templates map[string]ServiceTemplate) error {
+	if err := s.write(filepath.Join(s.ConfigRoot, name+".yaml"), templates); err != nil {
+		return err
+	}
+	if err := s.stage(templates); err != nil {
+		return err
+	}
+	return s.activate(templates)
 }
 
-func (b *ServiceBuilder) Remove(template *SBTemplate) error {
-	yamlPath := b.serviceYamlPath(template.name)
+// using the servicebuilder yaml files, find any extraneous runit services and
+// remove them
+// runsvdir automatically stops services that no longer exist, explicit stop is
+// not required
+func (s *ServiceBuilder) Prune() error {
+	configs, err := s.loadConfigDir()
+	if err != nil {
+		return err
+	}
 
-	err := os.Remove(yamlPath)
+	links, err := ioutil.ReadDir(s.RunitRoot)
 	if err != nil {
 		return err
 	}
+	for _, link := range links {
+		if _, exists := configs[link.Name()]; !exists {
+			if err = os.Remove(filepath.Join(s.RunitRoot, link.Name())); err != nil {
```

Comment:
Is this consistent with `servicebuilder`? I didn't realize it would remove *any* service under the runit root dir without it being provided in the configs.